### PR TITLE
Fix closure references in Daml Repl

### PR DESCRIPTION
**Pull request by cocreature**
Turns out the comment "we probably need to extend this to merge the
modules from each line" was exactly correct: If the result evaluates
to a closure (or a value including a closure), it can reference
definitions from the current module. This happens if the simplifier
lifted something out of the current definition (otherwise we have only
one and it cannot be recursive so no chance of leaking a reference).

This PR fixes this by checking whether the result references the
current module and if it does, we keep it around.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/iterable/SExprIterable.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.speedy.iterable
+
+import com.daml.lf.speedy.{SExpr, SValue}
+import scala.jdk.CollectionConverters._
+
+object SExprIterable {
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/iterable/SExprIterable.scala:9]: @remyhaemmerle-da I don’t like any of this but I don’t know how to do it nicer. If you have suggestions, I’m very open.

Comment (aherrmann-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/iterable/SExprIterable.scala:9]: @cocreature Maybe add a comment explaining how this iterates. As I understand it, it iterates only over the top layer but doesn't recurse into depth.

**Review by aherrmann-da — APPROVED** [review]

File: compiler/repl-service/server/src/main/scala/com/digitalasset/daml/lf/ReplServiceMain.scala [hunk 2]
```diff
@@ -177,6 +179,22 @@ object ReplService {
...
+          if (ref.packageId == homePackageId) {
+            acc + ref.modName
+          } else {
+            acc
+          }
+        case e => SExprIterable(e).foldLeft(acc)(moduleRefs)
+      }
+
+    SExprIterable(v).foldLeft(Set.empty[ModuleName])(moduleRefs)
+  }
```
Comment (remyhaemmerle-da) [inline compiler/repl-service/server/src/main/scala/com/digitalasset/daml/lf/ReplServiceMain.scala:197]: Can we move that to `com.daml.lf.speedy.SValue` and make the `SExprIterable` `speedy` private ?

**Review by remyhaemmerle-da — APPROVED** [review]
:cry:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/iterable/SExprIterable.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/iterable/SExprIterable.scala]: added a comment, yes your observation is exactly right. Works like our other iterables (e.g. `ExprIterable`) or Haskell’s plated stuff

File: compiler/repl-service/server/src/main/scala/com/digitalasset/daml/lf/ReplServiceMain.scala (hunk 2 above)
Comment (cocreature) [reply compiler/repl-service/server/src/main/scala/com/digitalasset/daml/lf/ReplServiceMain.scala:197]: I kept this function here but moved the code under `com.daml.lf.speedy.repl.ReplServiceMain` and made the other stuff `private`.

Comment (remyhaemmerle-da) [reply compiler/repl-service/server/src/main/scala/com/digitalasset/daml/lf/ReplServiceMain.scala:197]: Thanks.

**Pull request merged**